pkg/controller/core.oam.dev/v1alpha2/application/revision.go

```diff
@@ -291,3 +293,73 @@ func ComputeAppRevisionHash(appRevision *v1beta1.ApplicationRevision) (string, e
 	// compute the hash of the entire structure
 	return utils.ComputeSpecHash(&appRevisionHash)
 }
+
+// cleanUpApplicationRevision check all appRevisions of the application, remove them if the number of them exceed the limit
+func cleanUpApplicationRevision(ctx context.Context, h *appHandler) error {
+	listOpts := []client.ListOption{
+		client.InNamespace(h.app.Namespace),
+		client.MatchingLabels{oam.LabelAppName: h.app.Name},
+	}
+	appRevisionList := new(v1beta1.ApplicationRevisionList)
+	// controller-runtime will cache all appRevision by default, there is no need to watch or own appRevision in manager
+	if err := h.r.List(ctx, appRevisionList, listOpts...); err != nil {
+		return err
+	}
+	needKill := len(appRevisionList.Items) - h.r.appRevisionLimit
```

Comment:
```suggestion
	needKill := len(appRevisionList.Items) - h.r.appRevisionLimit - len(usingRevision)
```

We should keep all using revisions and extra 10 ones

Reply:
Got it